### PR TITLE
Consolidate common sub-expression evaluation logic

### PR DESCRIPTION
**Pull request by mbasmanova**
CSE evaluation logic in Expr.cpp was split between
checkUpdateSharedSubexprValues and checkGetSharedSubexprValues method an
interleaved with regular evaluation making it hard to following. This change
consolidates the logic in a single evaluateSharedSubexpr method. 

More refactoring is needed to make this logic easier to read and test. This 
is a first step.

Depends on #3900

### CODE REVIEW

**Review by xiaoxmeng — APPROVED** [review]
@mbasmanova LGTM. Thanks for the refactoring!

File: velox/expression/Expr.cpp [hunk 1]
```diff
@@ -567,96 +563,82 @@ void Expr::eval(
...
+  ScopedFinalSelectionSetter setter(
+      context, newFinalSelection, true /*checkCondition*/, true /*override*/);
+
+  evalEncodings(*missingRows, context, sharedSubexprValues_);
+
+  // Clear the rows which failed to compute.
+  context.deselectErrors(*missingRows);
+
+  sharedSubexprRows_->select(*missingRows);
+  context.moveOrCopyResult(sharedSubexprValues_, *missingRows, result);
```
Comment (bikramSingh91) [inline velox/expression/Expr.cpp:641]: The "missingRows" over here only copies missing rows and we lose out on the rows that were shared. instead it needs the full set of "rows" minus the rows that errored out which we remove in L638.
See #3958

Comment (mbasmanova) [reply velox/expression/Expr.cpp:641]: @bikramSingh91 Thank you for discovering this bug. It needs to be `context.moveOrCopyResult(sharedSubexprValues_, rows, result);` I think.
